bundles/remote_services/discovery_zeroconf/src/discovery_zeroconf_activator.c

```diff
@@ -18,43 +18,116 @@
  */
 #include "discovery_zeroconf_announcer.h"
 #include "discovery_zeroconf_watcher.h"
+#include "endpoint_listener.h"
+#include "remote_constants.h"
+#include "remote_service_admin.h"
 #include "celix_log_helper.h"
 #include "celix_bundle_activator.h"
-#include "celix_types.h"
+#include "celix_dm_component.h"
+#include "celix_dm_service_dependency.h"
+#include "celix_constants.h"
 #include "celix_errno.h"
+#include <stdio.h>
+#include <assert.h>
 
 typedef struct discovery_zeroconf_activator {
     celix_log_helper_t *logHelper;
     discovery_zeroconf_announcer_t *announcer;
     discovery_zeroconf_watcher_t  *watcher;
+    endpoint_listener_t endpointListener;
 }discovery_zeroconf_activator_t;
 
 celix_status_t discoveryZeroconfActivator_start(discovery_zeroconf_activator_t *act, celix_bundle_context_t *ctx) {
     celix_status_t status = CELIX_SUCCESS;
-    celix_autoptr(celix_log_helper_t) logger = celix_logHelper_create(ctx,"celix_rsa_zeroconf_discovery");
+    const char *fwUuid = celix_bundleContext_getProperty(ctx, CELIX_FRAMEWORK_UUID, NULL);
+    if (fwUuid == NULL) {
+        return CELIX_BUNDLE_EXCEPTION;
+    }
+    celix_autoptr(celix_log_helper_t) logger = celix_logHelper_create(ctx, "celix_rsa_zeroconf_discovery");
     if (logger == NULL) {
         return CELIX_ENOMEM;
     }
-    celix_autoptr(discovery_zeroconf_announcer_t) announcer = NULL;
-    status = discoveryZeroconfAnnouncer_create(ctx, logger, &announcer);
+    //init announcer
+    celix_autoptr(celix_dm_component_t) announcerCmp = celix_dmComponent_create(ctx, "DZC_ANNOUNCER_CMP");
+    if (announcerCmp == NULL) {
+        return CELIX_ENOMEM;
+    }
+    status = discoveryZeroconfAnnouncer_create(ctx, logger, &act->announcer);
     if (status != CELIX_SUCCESS) {
         return status;
     }
-    celix_autoptr(discovery_zeroconf_watcher_t) watcher = NULL;
-    status = discoveryZeroconfWatcher_create(ctx, logger, &watcher);
+    celix_dmComponent_setImplementation(announcerCmp, act->announcer);
+    CELIX_DM_COMPONENT_SET_IMPLEMENTATION_DESTROY_FUNCTION(announcerCmp, discovery_zeroconf_announcer_t, discoveryZeroconfAnnouncer_destroy);
+    celix_properties_t *props = celix_properties_create();
+    if (props == NULL) {
+        return CELIX_ENOMEM;
+    }
+    char scope[256] = {0};
+    (void)snprintf(scope, sizeof(scope), "(&(%s=*)(%s=%s))", CELIX_FRAMEWORK_SERVICE_NAME,
+                    OSGI_RSA_ENDPOINT_FRAMEWORK_UUID, fwUuid);
+    celix_properties_set(props, OSGI_ENDPOINT_LISTENER_SCOPE, scope);
+    celix_properties_set(props, "DISCOVERY", "true");//Only use to avoid the discovery calls to unnecessary endpoint listener service.Endpoint should be filtered by the scope.
```

Comment:
I am not sure if I get this. As also stated in the comment, I expected that the OSGI_ENDPOINT_LISTENER_SCOPE was enough. Is the "DISCOVERY" property added for additional performance optimization?

OSGi EndpointListener doc:
> Endpoint Listener services can express their scope with the service property [ENDPOINT_LISTENER_SCOPE](https://docs.osgi.org/javadoc/osgi.cmpn/8.1.0/org/osgi/service/remoteserviceadmin/EndpointListener.html#ENDPOINT_LISTENER_SCOPE). This service property is a list of filters. An Endpoint Description should only be given to a Endpoint Listener when there is at least one filter that matches the Endpoint Description properties. This filter model is quite flexible. For example, a discovery bundle is only interested in locally originating Endpoint Descriptions. The following filter ensure that it only sees local endpoints.
   (org.osgi.framework.uuid=72dc5fd9-5f8f-4f8f-9821-9ebb433a5b72)

Reply:
> Is the "DISCOVERY" property added for additional performance optimization?

yes, `ENDPOINT_LISTENER_SCOPE` is enough to ensure that the endpoint listener of discovery only sees local endpoint.

Reply:
Ok. I see this is used in discovery_common / discovery_endpointListenerAdded. I am not a fan of this approach, but IMO we can keep this for now.